### DISCUSSION TITLE
Programmatically prefetching

### DISCUSSION THREAD

**Discussion by ThewBear** [gatsbyjs/gatsby#20568] [Ideas / Feature Requests]
###  Summary

Currently gatsby-link prefetch route on `Link` component `componentDidMount`.
https://github.com/gatsbyjs/gatsby/blob/762506f714ca5c22764aa9befd3042f78b6d9d6a/packages/gatsby-link/src/index.js#L74-L79

It would be great if we also support some kind of prefetch for `navigate` function or a new programmatically prefetch page function. My current solution is using `window.___loader.enqueue` which is _technically unsupported_.

https://github.com/gatsbyjs/gatsby/blob/6f3eb9718330a1fc2debad44c1d0c7a92dbe94ea/docs/docs/production-app.md#L120

**Comment by karlhorky** [comment]
Super interesting @ThewBear!

I would also have a use case for it, from #20745. "These solutions" is talking about access control - denying access per-user and per-page on an Express static file server, which will cause failed requests with pre-fetching.

> One thing that these solutions cause is a lot of failed pre-fetching requests for users without full access:
> 
> ![Screen Shot 2020-01-21 at 12 52 12](https://user-images.githubusercontent.com/1935696/72802870-073fee00-3c4d-11ea-906e-2a8caad97f88.png)
> 
> Maybe there could be a way to configure pre-fetching client-side? So that different resources could be pre-fetched per user?

**Comment by karlhorky** [comment]
> My current solution is using `window.___loader.enqueue` which is *technically unsupported*.

@ThewBear Could you share your current solution with us?

I may use it in the interim to work around the failed requests in my proof of concept Gatsby + server-side Auth0 + Express repo: https://github.com/karlhorky/gatsby-serverside-auth0

**Comment by ThewBear** [comment]
@karlhorky I call `window.___loader.enqueue` inside useEffect hooks in Page component.

```jsx
import React, { useEffect } from "react";
import { navigate } from "gatsby";

const Page1 = () => {
  useEffect(() => {
    /* global ___loader */
    ___loader.enqueue("/page-2");
  }, []);

  return <button onClick={() => navigate("/page-2")}></button>
}
```

`gatsby-link` also implements this.
https://github.com/gatsbyjs/gatsby/blob/721e9d1cf4a53604cddb7901003046bd3649836c/packages/gatsby-link/src/index.js#L70

**Comment by karlhorky** [comment]
Ah interesting. I guess a new Gatsby `<Link />` component could be created with a `prefetch` prop, to control this behavior:

```jsx
<Link prefetch={isAllowed(user, page)} />
```

I'll have to see whether this is useful enough for us. If so, maybe I'll implement + publish this too.

**Comment by karlhorky** [comment]
Not sure why `not stale` was removed @LekoArts, this is still a valid issue.

**Comment by erpatterson11** [comment]
This is something that [Next.js implements](https://nextjs.org/docs/api-reference/next/router#routerprefetch) and would be really helpful to have in Gatsby.

**Comment by wahidshafique** [comment]
Well, its been over a year, are there still plans for some sort of programmatic prefetch like
```
router.prefetch(url, as)
```
